File: mobile/src/screens/SnapshotDetailScreen.js
import { useEffect, useState } from 'react';
import { View, Text, ScrollView, StyleSheet, ActivityIndicator } from 'react-native';
import { fetchJson } from '../api/client';

export default function SnapshotDetailScreen({ route }) {
  const { jobId, snapshotId } = route.params || {};
  const [snapshot, setSnapshot] = useState(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    if (!jobId || !snapshotId) {
      setLoading(false);
      return;
    }
    let cancelled = false;
    fetchJson(`/api/jobs/${jobId}/snapshots/${snapshotId}`)
      .then((data) => {
        if (!cancelled) setSnapshot(data);
      })
      .catch(() => {
        if (!cancelled) setSnapshot(null);
      })
      .finally(() => {
        if (!cancelled) setLoading(false);
      });
    return () => { cancelled = true; };
  }, [jobId, snapshotId]);

  if (loading) {
    return (
      <View style={styles.centered}>
        <ActivityIndicator size="large" />
      </View>
    );
  }

  if (!snapshot) {
    return (
      <View style={styles.centered}>
        <Text>Snapshot not found</Text>
      </View>
    );
  }

  let details = {};
  try {
    const rd = snapshot.route_details;
    details = typeof rd === 'string' ? JSON.parse(rd) : rd || {};
  } catch (_) {}
  const steps = details.steps || [];
  const points = details.points || [];

  return (
    <ScrollView style={styles.container}>
      <View style={styles.section}>
        <Text style={styles.label}>Collected at</Text>
        <Text style={styles.value}>{snapshot.collected_at ? new Date(snapshot.collected_at).toLocaleString() : '—'}</Text>
      </View>
      <View style={styles.section}>
        <Text style={styles.label}>Duration</Text>
        <Text style={styles.value}>{snapshot.duration_seconds != null ? `${Math.round(snapshot.duration_seconds / 60)} min` : '—'}</Text>
      </View>
      <View style={styles.section}>
        <Text style={styles.label}>Distance</Text>
        <Text style={styles.value}>{snapshot.distance_meters != null ? `${(snapshot.distance_meters / 1000).toFixed(2)} km` : '—'}</Text>
      </View>
      {points.length > 0 && (
        <View style={styles.section}>
          <Text style={styles.sectionTitle}>Route</Text>
          <Text style={styles.hint}>{points.length} points (map can be added with react-native-maps)</Text>
        </View>
      )}
      {steps.length > 0 && (
        <View style={styles.section}>
          <Text style={styles.sectionTitle}>Turn-by-turn</Text>
          {steps.map((step, i) => (
            <View key={i} style={styles.stepRow}>
              <Text style={styles.stepNum}>{i + 1}.</Text>
              <Text style={styles.stepText}>{step.html_instructions ? step.html_instructions.replace(/<[^>]*>/g, '').trim() : step.instruction || '—'}</Text>
            </View>
          ))}
        </View>
      )}
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#f5f5f5' },
  centered: { flex: 1, justifyContent: 'center', alignItems: 'center' },
  section: { backgroundColor: '#fff', marginHorizontal: 16, marginVertical: 8, padding: 16, borderRadius: 8 },
  sectionTitle: { fontSize: 16, fontWeight: '600', marginBottom: 12, color: '#111' },
  label: { fontSize: 12, color: '#666', marginBottom: 4 },
  value: { fontSize: 16, color: '#111' },
  hint: { fontSize: 14, color: '#666' },
  stepRow: { flexDirection: 'row', marginBottom: 8 },
  stepNum: { fontWeight: '600', marginRight: 8, color: '#374151' },
  stepText: { flex: 1, fontSize: 14, color: '#111' },
});

File: mobile/lib/screens/snapshot_detail_screen.dart
import 'package:flutter/material.dart';
import '../api/client.dart';

class SnapshotDetailScreen extends StatefulWidget {
  const SnapshotDetailScreen({
    super.key,
    required this.jobId,
    required this.snapshotId,
  });

  final String jobId;
  final String snapshotId;

  @override
  State<SnapshotDetailScreen> createState() => _SnapshotDetailScreenState();
}

class _SnapshotDetailScreenState extends State<SnapshotDetailScreen> {
  Map<String, dynamic>? _snapshot;
  bool _loading = true;
  String? _error;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final data = await apiClient.get(
        '/api/jobs/${widget.jobId}/snapshots/${widget.snapshotId}',
      );
      if (mounted) setState(() {
            _snapshot = data;
            _loading = false;
          });
    } catch (e) {
      if (mounted) setState(() {
            _error = e.toString().replaceFirst('Exception: ', '');
            _loading = false;
          });
    }
  }

  @override
  Widget build(BuildContext context) {
    if (_loading) {
      return Scaffold(
        appBar: AppBar(title: const Text('Snapshot')),
        body: const Center(child: CircularProgressIndicator()),
      );
    }
    if (_error != null) {
      return Scaffold(
        appBar: AppBar(title: const Text('Snapshot')),
        body: Center(
          child: Padding(
            padding: const EdgeInsets.all(24),
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Text(_error!, textAlign: TextAlign.center),
                const SizedBox(height: 16),
                FilledButton(onPressed: _load, child: const Text('Retry')),
              ],
            ),
          ),
        ),
      );
    }
    final s = _snapshot!;
    final duration = s['duration_minutes'];
    final at = s['collected_at'] as String? ?? '';
    final steps = s['route_details']?['steps'] as List<dynamic>? ?? [];

    return Scaffold(
      appBar: AppBar(title: Text(duration != null ? '$duration min' : 'Snapshot')),
      body: ListView(
        padding: const EdgeInsets.all(16),
        children: [
          if (at.isNotEmpty) Text('Collected: $at'),
          const SizedBox(height: 16),
          const Text('Steps', style: TextStyle(fontWeight: FontWeight.bold)),
          const SizedBox(height: 8),
          ...steps.map<Widget>((e) {
            final step = e as Map<String, dynamic>;
            final instr = step['instruction'] as String? ?? '';
            final dist = step['distance_text'] as String?;
            return Padding(
              padding: const EdgeInsets.only(bottom: 8),
              child: Row(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Icon(Icons.directions, size: 20),
                  const SizedBox(width: 8),
                  Expanded(
                    child: Text(
                      instr,
                      style: const TextStyle(fontSize: 14),
                    ),
                  ),
                  if (dist != null) Text(dist, style: const TextStyle(fontSize: 12)),
                ],
              ),
            );
          }),
        ],
      ),
    );
  }
}
